plugins/docker/docker_stats_input.go

```diff
@@ -0,0 +1,229 @@
+
+	di.conf = config.(*DockerStatsInputConfig)
+	di.stopChan = make(chan error)
+	di.closer = make(chan struct{})
+	di.statsstream = make(chan *DockerStat)
+	di.attachErrors = make(chan error)
+
+	m, err := NewStatsManager(di.conf.Endpoint, di.conf.CertPath, di.attachErrors, di.conf.NameFromEnv, di.conf.FieldsFromEnv, di.conf.FieldsFromLabels)
+	if err != nil {
+		return fmt.Errorf("DockerStatsInput: failed to attach: %s", err.Error())
+	}
+
+	di.statsMgr = m
+	return nil
+}
+
+func (di *DockerStatsInput) Run(ir pipeline.InputRunner, h pipeline.PluginHelper) error {
+	var ok   bool
+	di.statsMgr.ir = ir
+	hostname := h.Hostname()
+
+	go di.statsMgr.Run(di.statsstream, di.closer, di.stopChan)
+
+	err := withRetries(func() error { return di.statsMgr.client.AddEventListener(di.statsMgr.events) })
+	// Get the InputRunner's chan to receive empty PipelinePacks
+	packSupply := ir.InChan()
+
+	ok = true
+	for ok {
+		select {
+		case statsline := <-di.statsstream:
+			di.pack = <-packSupply
+
+			di.pack.Message.SetType("DockerStats")
+			di.pack.Message.SetLogger(statsline.Container)
+			di.pack.Message.SetHostname(hostname)   // Use the host's hosntame
+			di.pack.Message.SetPayload(statsline.StatsString)
+			di.pack.Message.SetTimestamp(statsline.Time.UnixNano())
+			di.pack.Message.SetUuid(uuid.NewRandom())
+			di.MarshalStatsFields("stat", reflect.ValueOf(statsline.Stat), ir)
+
+			for name, value := range statsline.Fields {
+				field, err := message.NewField(name, value, "")
+				if err != nil {
+					ir.LogError(
+						fmt.Errorf("can't add '%s' field: %s", name, err.Error()),
+					)
+					continue
+				}
+
+				di.pack.Message.AddField(field)
+			}
+			ir.Deliver(di.pack)
+
+
+		case err, ok = <-di.attachErrors:
+			if !ok {
+				err = errors.New("Docker event channel closed")
+				break
+			}
+			ir.LogError(fmt.Errorf("Attacher error: %s", err))
+
+		case err = <-di.stopChan:
+			ok = false
+		}
+	}
+
+	di.closer <- struct{}{}
+	close(di.statsstream)
+	if err != nil {
+		ir.LogError(err)
+		return errors.New(
+			"Failed to attach to Docker containers after retrying. Plugin giving up.")
+	}
+	if err != nil {
+		ir.LogError(err)
+		return errors.New(
+			"Failed to add Docker event listener after retrying. Plugin giving up.")
+	}
+	return nil
+}
+
+func (di *DockerStatsInput) CleanupForRestart() {
+	// Intentionally left empty. Cleanup happens in Run()
+}
+
+func (di *DockerStatsInput) Stop() {
+	close(di.stopChan)
+}
+
+func init() {
+	pipeline.RegisterPlugin("DockerStatsInput", func() interface{} {
+		return new(DockerStatsInput)
+	})
+}
+
+
+// This method marshals the fields of a Stats struct into a map[string]string fields object
+// Core logic for handling reflect.value kinds based on
+// https://github.com/golang/go/blob/dbaf5010b33e5819050ebb0a387eb0bff2cfb8bf/src/fmt/scan.go#L994
+func (di *DockerStatsInput) MarshalStatsFields(path string, v reflect.Value, ir pipeline.InputRunner) {
```

Comment:
I was suggesting that you can probably let `fmt` do some of this for you. e.g. https://play.golang.org/p/wD-cGK-n2g